### PR TITLE
Api

### PR DESCRIPTION
**Pull request by water-korra**

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,162 +1,162 @@
-{
```
Comment (akaeyuhi) [inline .eslintrc.json:1]: zachem

File: .prettierrc [hunk 2]
```diff
@@ -1,10 +1,10 @@
-{
```
Comment (akaeyuhi) [inline .prettierrc:1]: zachem x2

File: src/api/Controller.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+const db = require('../database/db')
```
Comment (akaeyuhi) [inline src/api/Controller.js:1]: Use prettier pls

File: src/api/Controller.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    async getRecipeById(req,res) {
+        try {
+            const {id} = req.body
+            const neededRecipe = await db.getRecipe(id)
+            return res.json({message: neededRecipe})
+        } catch (e) {
+            console.log(e)
+            res.status(400).json({message: 'error by getting recipe by id'});
```
Comment (akaeyuhi) [inline src/api/Controller.js:29]: If you are using express.json middleware .json is redundant

File: src/api/authController.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+class authController {
+    async registration(req,res) {
+        try {
+            const errors = validationResult(req)
+            if(!errors.isEmpty()) {
+                return res.status(400).json({message: "Registration mistake",errors})
+            }
+              // const candidate = db.getUserData(username)
```
Comment (akaeyuhi) [inline src/api/authController.js:23]: Delete commented lines

File: src/api/authController.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        } catch (e) {
+            console.log(e)
+            res.status(400).json({message: 'Registration error'});
+        }
+    }
+
+    async login(req,res) {
+        try {
+            const {username,password} = req.body
+            const user =  await db.getUserData(username)
```
Comment (akaeyuhi) [inline src/api/authController.js:41]: Prettier

File: src/api/authController.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            // }
+            const token = generateAccessToken(user.id)
+            // return res.json({token})
+            return res.json(`Hello ${username}`)
+
+        }
+
+
+        catch (e) {
+            console.log(e)
```
Comment (akaeyuhi) [inline src/api/authController.js:67]: Delete console.log() here and further

Comment (akaeyuhi) [reply src/api/authController.js:67]: Using log for error is a bad practice

File: src/api/authconfig.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+    secret: "keykey"
```
Comment (akaeyuhi) [inline src/api/authconfig.js:2]: Generate random secret on production

File: src/api/package.json [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  "dependencies": {
+    "bcryptjs": "^2.4.3",
+    "cookie-parser": "^1.4.6",
+    "cors": "^2.8.5",
+    "dotenv": "^10.0.0",
+    "express": "^4.17.2",
+    "express-validator": "^6.14.0",
+    "fastify": "^3.25.3",
+    "mongoose": "^6.1.4",
```
Comment (akaeyuhi) [inline src/api/package.json:23]: Unysed dependencies

File: src/database/db.js [hunk 10]
```diff
@@ -89,9 +89,62 @@ async function addUserFavRecipeToDb(userId, recipeId) {
...
+  const getRecipesRequest = `SELECT * FROM recipes`;
+  const res = await sqlRequest(getRecipesRequest);
+  return res; //returns array of objects(recipes)
+}
+
+async function getUserData(username) {
+  const getUserDataRequest = `SELECT * FROM usersdata WHERE username = ?`;
+  const res = await sqlRequest(getUserDataRequest, [username]);
+  if (res[0]) return res[0];
+  return undefined;
```
Comment (akaeyuhi) [inline src/database/db.js:102]: Dont return undefined, it`s returned by default

File: src/database/db.js [hunk 11]
```diff
@@ -89,9 +89,62 @@ async function addUserFavRecipeToDb(userId, recipeId) {
...
+  const getUserFavRecipesRequest = `SELECT * FROM favourite_recipes 
+  WHERE user_id = ?`;
+  const res = await sqlRequest(getUserFavRecipesRequest, [userId]);
+  if (res[0]) {
+    for (const el of res) {
+      recipesArr.push(el.recipe_id);
+    }
+    return recipesArr; //retuns array of recipes ids
+  }
+  return undefined;
```
Comment (akaeyuhi) [inline src/database/db.js:116]: and here

File: src/database/db.js [hunk 12]
```diff
@@ -89,9 +89,62 @@ async function addUserFavRecipeToDb(userId, recipeId) {
...
+}
+
+async function getRecipe(recipeId) {
+  const getRecipeRequest = `SELECT * FROM recipes WHERE id = ?`;
+  const res = await sqlRequest(getRecipeRequest, [recipeId]);
+  if (res[0]) return res[0];
+  return undefined;
```
Comment (akaeyuhi) [inline src/database/db.js:123]: and here

**Review by akaeyuhi — CHANGES_REQUESTED** [review]
Huge code style problems

**Review by akaeyuhi — COMMENTED** [review]
Delete DS store files

File: src/api/controllers/Controller.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    console.log(e);
+    res
+      .status(400)
+      .json({ message: 'Error ocured while getting recipes list' });
+  }
+
+  async addUserRecipe(req, res) {
+    try {
+      const { recipe_id } = req.body;
+      const user_id = tokenService.getUserId(req);
```
Comment (akaeyuhi) [inline src/api/controllers/Controller.js:59]: camelcase pls

File: src/api/controllers/Controller.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+      res.status(400).json({ message: 'Error occured while adding recipe' });
+    }
+  }
+
+  async getUserRecipes(req, res) {
+    try {
+      const user_id = tokenService.getUserId(req);
+      const userFavId = await db.getUserFavRecipes(user_id);
+      let userFavList = [];
+      for (let i = 0; i < userFavId.length; i++) {
```
Comment (akaeyuhi) [inline src/api/controllers/Controller.js:73]: for of would be better

File: src/api/controllers/Controller.js [hunk 15]
```diff
@@ -0,0 +1,97 @@
+const db = require('../../database/db');
+const jwt = require('jsonwebtoken');
+const tokenService = require('../service/token-service');
+const ParseService = require('../service/recipe-service');
+// const fs = require('fs')
+
+
+let jsonData = require('../array.json')
+ParseService.addRecipesFromParcer(jsonData)
```
Comment (akaeyuhi) [inline src/api/controllers/Controller.js:9]: Is this a class of smth?

File: src/api/controllers/authController.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+const db = require('../../database/db');
+const { validationResult } = require('express-validator');
+const tokenService = require('../service/token-service');
+
+class authController {
```
Comment (akaeyuhi) [inline src/api/controllers/authController.js:7]: Name classes with capital letter

File: src/api/service/recipe-service.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
+const db = require('../../database/db');
+
+class ParseService {
+    addRecipesFromParcer = (jsonData) => {
```
Comment (akaeyuhi) [inline src/api/service/recipe-service.js:4]: Arrow function in a class is weird

File: src/api/service/recipe-service.js [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+        for(let i = 0;i<jsonData.length;i++){
```
Comment (akaeyuhi) [inline src/api/service/recipe-service.js:5]: code style and for of

File: src/api/service/recipe-service.js [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+            let name = jsonData[i].name
+            let category_id = jsonData[i].category_id
+            let products = jsonData[i].products
+            let description = jsonData[i].description
+            let image_link = jsonData[i].image_link
+            console.log(`name${i+1} = ${name}`);
+            db.addRecipeToDb(name,category_id,products,description,image_link)
```
Comment (akaeyuhi) [inline src/api/service/recipe-service.js:12]: You don`t need a whole class just to create an alias for db api function

File: src/database/db.js [hunk 20]
```diff
@@ -89,9 +89,89 @@ async function addUserFavRecipeToDb(userId, recipeId) {
...
+  const res = await sqlRequest(deleteFavRecipeRequest, [userId, recipeId]);
+  if (res.affectedRows === 1) return true;
+  return false;
+}
+
+async function getCategoryName(categoryId) {
+  const CategoryNameRequest = `SELECT name FROM categories WHERE id = ?`;
+  const res = await sqlRequest(CategoryNameRequest, [categoryId]);
+  if (res[0]) return res[0].name;
+  return undefined;
```
Comment (akaeyuhi) [inline src/database/db.js:152]: Undefined once again

File: src/database/db.js [hunk 21]
```diff
@@ -89,9 +89,89 @@ async function addUserFavRecipeToDb(userId, recipeId) {
...
+  const deleteFavRecipeRequest = `DELETE FROM favourite_recipes 
+  WHERE user_id = ?
+  AND recipe_id = ?`;
+  const res = await sqlRequest(deleteFavRecipeRequest, [userId, recipeId]);
+  if (res.affectedRows === 1) return true;
+  return false;
+}
+
+async function getCategoryName(categoryId) {
+  const CategoryNameRequest = `SELECT name FROM categories WHERE id = ?`;
```
Comment (akaeyuhi) [inline src/database/db.js:149]: Capital letter and not a class

File: src/database/db.js [hunk 22]
```diff
@@ -89,9 +89,89 @@ async function addUserFavRecipeToDb(userId, recipeId) {
...
+  if (res[0]) return res[0].name;
+  return undefined;
+}
+
+async function findRecipes(str) {
+  const findRecipesRequest = `SELECT * FROM recipes WHERE name LIKE ?`;
+  const res = await sqlRequest(findRecipesRequest, [`%${str}%`]);
+  console.dir(res)
+  if (res[0]) return res;
+  return undefined;
```
Comment (akaeyuhi) [inline src/database/db.js:160]: And again

**Review by akaeyuhi — CHANGES_REQUESTED** [review]
Still a lot of things to refactor here

File: src/api/controllers/Controller.js [hunk 23]
```diff
@@ -1,32 +1,21 @@
...
-const jwt = require('jsonwebtoken');
 const tokenService = require('../service/token-service');
 const ParseService = require('../service/recipe-service');
-// const fs = require('fs')
-
-
-let jsonData = require('../array.json')
-ParseService.addRecipesFromParcer(jsonData)
 
+// let jsonData = require('../../parser/recipes.json');
```
Comment (akaeyuhi) [inline src/api/controllers/Controller.js:5]: ирина *б твою мть

File: src/api/routes/authRouter.js [hunk 24]
```diff
@@ -1,20 +1,16 @@
...
-    check('password', 'Password should be longer than 6 symbols').isLength({
-      min: 4,
-      max: 14,
-    }),
-  ],
+  userService.validate,
   authcontroller.registration
 );
-authrouter.post('/login', authcontroller.login);
+
```
Comment (akaeyuhi) [inline src/api/routes/authRouter.js:11]: ок

**Review by akaeyuhi — CHANGES_REQUESTED** [review]
Vesma neudovletvoritelno, zaimetes` ulutscheniem

File: src/api/service/recipe-service.js (hunk 17 above)
Comment (water-korra) [reply src/api/service/recipe-service.js:25]: fixed cap

File: src/api/controllers/authController.js [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+class AuthController {
+  async registration(req, res) {
+    try {
+      const errors = validationResult(req);
+      if (!errors.isEmpty()) {
+        return res
+          .status(400)
+          .json({ message: 'Registration mistake', errors });
+      }
+      const { username, email, password } = req.body;
```
Comment (akaeyuhi) [inline src/api/controllers/authController.js:16]: better to check body to not get undefined but ok

**Review by akaeyuhi — APPROVED** [review]

**Pull request merged**